### PR TITLE
Refactor to use http.Request with context, other minor refactoring

### PR DESCRIPTION
**Pull request by johejo**
## Description

The main content of this PR is refactoring.
There should be little change in the logic.
Adjusted the interface so that the http handler context is properly propagated to http.Request for request to providers.

Also I found some dangerous type assertions and error handling, so I enabled errcheck linter and fixed some of them and including some other minor refactorings.

## Motivation and Context

Context support is necessary considering future support for OpenTelemetry tracer.

## How Has This Been Tested?

`make test`

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -130,13 +130,13 @@ func (l *Logger) Output(calldepth int, message string) {
...
 
-	l.stdLogTemplate.Execute(l.writer, stdLogMessageData{
+	_ = l.stdLogTemplate.Execute(l.writer, stdLogMessageData{
 		Timestamp: FormatTimestamp(now),
 		File:      file,
 		Message:   message,
 	})
 
-	l.writer.Write([]byte("\n"))
+	fmt.Fprint(l.writer, "\n")
```
Comment (JoelSpeed) [inline pkg/logger/logger.go:139]: Same as above

File: pkg/requests/requests_test.go [hunk 2]
```diff
@@ -86,23 +87,25 @@ func TestRequestUnparsedResponseUsingAccessTokenParameter(t *testing.T) {
 		}))
 	defer backend.Close()
 
-	response, err := RequestUnparsedResponse(
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline pkg/requests/requests_test.go:90]: Use Background in tests as TODO suggests we are going to replace it at some point in the future
```suggestion
	ctx := context.Background()
```

File: pkg/requests/requests_test.go [hunk 3]
```diff
@@ -86,23 +87,25 @@ func TestRequestUnparsedResponseUsingAccessTokenParameter(t *testing.T) {
...
 }
 
 func TestRequestUnparsedResponseUsingAccessTokenParameterFailedResponse(t *testing.T) {
-	backend := testBackend(t, 200, "some payload")
+	backend := testBackend(t, http.StatusOK, "some payload")
 	// Close the backend now to force a request failure.
 	backend.Close()
 
-	response, err := RequestUnparsedResponse(
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline pkg/requests/requests_test.go:107]:
```suggestion
	ctx := context.Background()
```

File: pkg/requests/requests_test.go [hunk 4]
```diff
@@ -121,13 +124,14 @@ func TestRequestUnparsedResponseUsingHeaders(t *testing.T) {
 		}))
 	defer backend.Close()
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline pkg/requests/requests_test.go:127]:
```suggestion
	ctx := context.Background()
```

File: providers/azure.go [hunk 5]
```diff
@@ -11,6 +12,7 @@ import (
 	"time"
 
 	"github.com/bitly/go-simplejson"
+
```
Comment (JoelSpeed) [inline providers/azure.go:15]: New line?

File: providers/azure.go [hunk 6]
```diff
@@ -120,70 +119,67 @@ func (p *AzureProvider) Redeem(redirectURL, code string) (s *sessions.SessionSta
...
 }
 
 func getEmailFromJSON(json *simplejson.Json) (string, error) {
-	var email string
-	var err error
-
-	email, err = json.Get("mail").String()
-
+	email, err := json.Get("mail").String()
 	if err != nil || email == "" {
```
Comment (JoelSpeed) [inline providers/azure.go:143]: If you were to reverse this, you could return early here and then do the rest of the method under lesser indent
```
	if err == nil && email != "" {
            return email, nil
        }
```

File: providers/azure_test.go [hunk 7]
```diff
@@ -132,8 +133,9 @@ func TestAzureProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testAzureProvider(bURL.Host)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/azure_test.go:136]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 8]
```diff
@@ -145,8 +147,9 @@ func TestAzureProviderGetEmailAddressMailNull(t *testing.T) {
...
```
Comment (JoelSpeed) [inline providers/azure_test.go:150]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 9]
```diff
@@ -158,8 +161,9 @@ func TestAzureProviderGetEmailAddressGetUserPrincipalName(t *testing.T) {
...
```
Comment (JoelSpeed) [inline providers/azure_test.go:164]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 10]
```diff
@@ -171,8 +175,9 @@ func TestAzureProviderGetEmailAddressFailToGetEmailAddress(t *testing.T) {
...
```
Comment (JoelSpeed) [inline providers/azure_test.go:178]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 11]
```diff
@@ -184,8 +189,9 @@ func TestAzureProviderGetEmailAddressEmptyUserPrincipalName(t *testing.T) {
...
```
Comment (JoelSpeed) [inline providers/azure_test.go:192]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 12]
```diff
@@ -197,8 +203,9 @@ func TestAzureProviderGetEmailAddressIncorrectOtherMails(t *testing.T) {
...
```
Comment (JoelSpeed) [inline providers/azure_test.go:206]:
```suggestion
	ctx := context.Background()
```

File: providers/azure_test.go [hunk 13]
```diff
@@ -209,10 +216,11 @@ func TestAzureProviderRedeemReturnsIdToken(t *testing.T) {
 	timestamp, err := time.Parse(time.RFC3339, "2006-01-02T22:04:05Z")
 	assert.Equal(t, nil, err)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/azure_test.go:219]:
```suggestion
	ctx := context.Background()
```

File: providers/bitbucket_test.go [hunk 14]
```diff
@@ -119,8 +120,9 @@ func TestBitbucketProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testBitbucketProvider(bURL.Host, "", "")
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/bitbucket_test.go:123]: Please change the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/digitalocean_test.go [hunk 15]
```diff
@@ -98,8 +100,9 @@ func TestDigitalOceanProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testDigitalOceanProvider(bURL.Host)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/digitalocean_test.go:103]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/facebook.go [hunk 16]
```diff
@@ -55,11 +56,11 @@ func getFacebookHeader(accessToken string) http.Header {
 }
 
 // GetEmailAddress returns the Account email address
-func (p *FacebookProvider) GetEmailAddress(s *sessions.SessionState) (string, error) {
+func (p *FacebookProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
 	if s.AccessToken == "" {
 		return "", errors.New("missing access token")
 	}
-	req, err := http.NewRequest("GET", p.ProfileURL.String()+"?fields=name,email", nil)
+	req, err := http.NewRequest(http.MethodGet, p.ProfileURL.String()+"?fields=name,email", nil)
```
Comment (JoelSpeed) [inline providers/facebook.go:63]:
```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.ProfileURL.String()+"?fields=name,email", nil)
```

File: providers/github.go [hunk 17]
```diff
@@ -135,7 +139,9 @@ func (p *GitHubProvider) hasOrg(accessToken string) (bool, error) {
 	return false, nil
 }
 
-func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+var rep1 = regexp.MustCompile(`(?s).*\<https://api.github.com/user/teams\?page=(.)&per_page=[0-9]+\>; rel="last".*`)
```
Comment (JoelSpeed) [inline providers/github.go:142]: Please move to the top of the file

File: providers/github_test.go [hunk 18]
```diff
@@ -104,8 +105,9 @@ func TestGitHubProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testGitHubProvider(bURL.Host)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/github_test.go:108]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/gitlab_test.go [hunk 19]
```diff
@@ -62,8 +63,9 @@ func TestGitLabProviderBadToken(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testGitLabProvider(bURL.Host)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/gitlab_test.go:66]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/google_test.go [hunk 20]
```diff
@@ -102,7 +102,8 @@ func TestGoogleProviderGetEmailAddress(t *testing.T) {
 	p.RedeemURL, server = newRedeemServer(body)
 	defer server.Close()
 
-	session, err := p.Redeem("http://redirect/", "code1234")
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/google_test.go:105]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/internal_util_test.go [hunk 21]
```diff
@@ -87,43 +91,49 @@ func (vtTest *ValidateSessionStateTest) Close() {
 func TestValidateSessionStateValidToken(t *testing.T) {
 	vtTest := NewValidateSessionStateTest()
 	defer vtTest.Close()
-	assert.Equal(t, true, validateToken(vtTest.provider, "foobar", nil))
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/internal_util_test.go:94]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/keycloak.go [hunk 22]
```diff
@@ -47,14 +48,13 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
 }
 
-func (p *KeycloakProvider) GetEmailAddress(s *sessions.SessionState) (string, error) {
+func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
 
-	req, err := http.NewRequest("GET", p.ValidateURL.String(), nil)
-	req.Header.Set("Authorization", "Bearer "+s.AccessToken)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.ValidateURL.String(), nil)
 	if err != nil {
-		logger.Printf("failed building request %s", err)
```
Comment (JoelSpeed) [inline providers/keycloak.go:55]: Should this be kept?

File: providers/keycloak_test.go [hunk 23]
```diff
@@ -98,8 +99,9 @@ func TestKeycloakProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testKeycloakProvider(bURL.Host, "")
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/keycloak_test.go:102]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/linkedin.go [hunk 24]
```diff
@@ -51,11 +52,11 @@ func getLinkedInHeader(accessToken string) http.Header {
 }
 
 // GetEmailAddress returns the Account email address
-func (p *LinkedInProvider) GetEmailAddress(s *sessions.SessionState) (string, error) {
+func (p *LinkedInProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
 	if s.AccessToken == "" {
 		return "", errors.New("missing access token")
 	}
-	req, err := http.NewRequest("GET", p.ProfileURL.String()+"?format=json", nil)
+	req, err := http.NewRequest(http.MethodGet, p.ProfileURL.String()+"?format=json", nil)
```
Comment (JoelSpeed) [inline providers/linkedin.go:59]:
```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.ProfileURL.String()+"?format=json", nil)
```

File: providers/linkedin_test.go [hunk 25]
```diff
@@ -98,8 +99,9 @@ func TestLinkedInProviderGetEmailAddress(t *testing.T) {
 	bURL, _ := url.Parse(b.URL)
 	p := testLinkedInProvider(bURL.Host)
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/linkedin_test.go:102]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/logingov.go [hunk 26]
```diff
@@ -88,66 +90,71 @@ type loginGovCustomClaims struct {
...
 		var pubkeys jose.JSONWebKeySet
-		myerr = json.Unmarshal(body, &pubkeys)
-		if myerr != nil {
-			return nil, myerr
+		err = json.Unmarshal(body, &pubkeys)
+		if err != nil {
+			return nil, err
+		}
+		if len(pubkeys.Keys) == 0 {
+			return nil, fmt.Errorf("no keys in jwt: pubkeys=%v", pubkeys)
```
Comment (JoelSpeed) [inline providers/logingov.go:118]:
```suggestion
			return nil, fmt.Errorf("no keys in jwks")
```

File: providers/logingov.go [hunk 27]
```diff
@@ -88,66 +90,71 @@ type loginGovCustomClaims struct {
...
 	})
 	if err != nil {
-		return
+		return err
 	}
 
-	claims := token.Claims.(*loginGovCustomClaims)
+	claims, ok := token.Claims.(*loginGovCustomClaims)
+	if !ok {
+		return fmt.Errorf("type assertion of toke.Claims failed: token=%s", token.Raw)
```
Comment (JoelSpeed) [inline providers/logingov.go:130]:
```suggestion
		return fmt.Errorf("type assertion of token.Claims failed: token=%s", token.Raw)
```

File: providers/logingov_test.go [hunk 28]
```diff
@@ -189,7 +190,8 @@ func TestLoginGovProviderSessionData(t *testing.T) {
 	p.PubJWKURL, pubjwkserver = newLoginGovServer(pubjwkbody)
 	defer pubjwkserver.Close()
 
-	session, err := p.Redeem("http://redirect/", "code1234")
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/logingov_test.go:193]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/nextcloud_test.go [hunk 29]
```diff
@@ -96,8 +98,9 @@ func TestNextcloudProviderGetEmailAddress(t *testing.T) {
 	p.ValidateURL.Path = userPath
 	p.ValidateURL.RawQuery = formatJSON
 
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/nextcloud_test.go:101]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/oidc_test.go [hunk 30]
```diff
@@ -156,7 +156,8 @@ func TestOIDCProviderRedeem(t *testing.T) {
 	server, provider := newTestSetup(body)
 	defer server.Close()
 
-	session, err := provider.Redeem(provider.RedeemURL.String(), "code1234")
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/oidc_test.go:159]: Please do the others in this file too
```suggestion
	ctx := context.Background()
```

File: providers/provider_default.go [hunk 31]
```diff
@@ -35,53 +39,50 @@ func (p *ProviderData) Redeem(redirectURL, code string) (s *sessions.SessionStat
...
 
 	// blindly try json and x-www-form-urlencoded
 	var jsonResponse struct {
 		AccessToken string `json:"access_token"`
 	}
 	err = json.Unmarshal(body, &jsonResponse)
-	if err == nil {
-		s = &sessions.SessionState{
-			AccessToken: jsonResponse.AccessToken,
+	if err != nil {
```
Comment (JoelSpeed) [inline providers/provider_default.go:70]: This is neater the other way around in my opinion

File: providers/provider_default_test.go [hunk 32]
```diff
@@ -1,16 +1,19 @@
...
-	"github.com/oauth2-proxy/oauth2-proxy/pkg/apis/sessions"
 	"github.com/stretchr/testify/assert"
+
+	"github.com/oauth2-proxy/oauth2-proxy/pkg/apis/sessions"
 )
 
 func TestRefresh(t *testing.T) {
 	p := &ProviderData{}
-	refreshed, err := p.RefreshSessionIfNeeded(&sessions.SessionState{
+	ctx := context.TODO()
```
Comment (JoelSpeed) [inline providers/provider_default_test.go:15]:
```suggestion
	ctx := context.Background()
```

File: watcher.go [hunk 33]
```diff
@@ -55,7 +55,7 @@ func WatchForUpdates(filename string, done <-chan bool, action func()) {
 				// can't be opened.
 				if event.Op&(fsnotify.Remove|fsnotify.Rename|fsnotify.Chmod) != 0 {
 					logger.Printf("watching interrupted on event: %s", event)
-					watcher.Remove(filename)
+					_ = watcher.Remove(filename)
```
Comment (JoelSpeed) [inline watcher.go:58]: Should this error not be checked?

File: providers/logingov.go (hunk 26 above)
Comment (johejo) [reply providers/logingov.go:118]: `jwks` How do you mean?
`jwts`?, `jwt keys` ?

**Comment by johejo** [comment]
It is my fault to have made many small wrong code. Thank you for your reviewing.
I will fix.

The import grouping fixes are effect of an goimports option.
https://github.com/golang/tools/blob/5d8e1897c761b13642bb6a477c03c1c1bf9f34e2/cmd/goimports/goimports.go#L61

File: oauthproxy.go [hunk 34]
```diff
@@ -340,34 +341,37 @@ func (p *OAuthProxy) displayCustomLoginForm() bool {
...
 	if err != nil {
-		return
+		return nil, err
 	}
 
 	if s.Email == "" {
-		s.Email, err = p.provider.GetEmailAddress(s)
+		s.Email, err = p.provider.GetEmailAddress(ctx, s)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (johejo) [reply oauthproxy.go]: I switched to logging considering the side effect.

File: watcher.go [hunk 35]
```diff
@@ -55,7 +55,9 @@ func WatchForUpdates(filename string, done <-chan bool, action func()) {
 				// can't be opened.
 				if event.Op&(fsnotify.Remove|fsnotify.Rename|fsnotify.Chmod) != 0 {
 					logger.Printf("watching interrupted on event: %s", event)
-					watcher.Remove(filename)
+					if err := watcher.Remove(filename); err != nil {
+						logger.Printf("failed to remove %s", filename)
```
Comment (JoelSpeed) [inline watcher.go:59]:
```suggestion
						logger.Printf("failed to remove watch on file %q", filename)
```

**Comment by johejo** [comment]
This PR has become too big.
I will recreate the PR focusing only on the changes to support context.

**Pull request closed without merging**